// player/lib/player.js
/**
 * Created by Knaufux on 11/23/2015.
 */

var random = require('../../../node_modules/random/lib/random');

module.exports = function(playersDB, id, attacksDB){
    var root = this;
    var DIRECTION = [
        [0,1],
        [-1,0],
        [1,0],
        [0,-1]
    ];

    this.connected = true;
    this.id = id;
    this.attacks = attacksDB;
    this.players = playersDB;
    this.db = playersDB.get(id);
    this.getServerData = function(){
        return {
            xPlayer: root.db.x,
            yPlayer: root.db.y,
            dirPlayer: root.db.dir,
            hlthPlayer: root.db.currHp,
            stmnPlayer: root.db.currentStm,
            xpPlayer: root.db.currXp,
            idPlayer: root.id,
            spritePlayer: root.db.sprite,
            chrgdTmPlayer: root.db.chargedTime + root.db.perks.chargedTimeMod
        }
    };
    this.attack = function(fxArr){
        root.db.currentStm -= 1;

        var atckPerks = root.attacks.get(root.db.attack);
        atckPerks.x = root.db.x + DIRECTION[root.db.dir][0] * 30 + random.randomIntRange(-1 * atckPerks.randomizePos, atckPerks.randomizePos);
        atckPerks.y = root.db.y + DIRECTION[root.db.dir][1] * 30 + random.randomIntRange(-1 * atckPerks.randomizePos, atckPerks.randomizePos);
        atckPerks.x += (32 - atckPerks.baseAoE) / 2;
        atckPerks.y += (32 - atckPerks.baseAoE) / 2;
        atckPerks.width = atckPerks.baseAoE;
        atckPerks.height = atckPerks.baseAoE;
        atckPerks.plrDmg = root.db.strength + root.db.perks.damage;
        atckPerks.creator = root.db;

        fxArr.push({
            x: atckPerks.x - (32 - atckPerks.baseAoE) / 2,
            y: atckPerks.y - (32 - atckPerks.baseAoE) / 2,
            graphic: atckPerks.graphic
        });
        root.players.update(root.db);

        return atckPerks;
    };
    this.levelUp = function(lvlArr){
        while(root.db.currXp >= root.db.maxXp){
            root.db.currXp -= root.db.maxXp;
            root.db.maxXp *= 2;
            root.db.level += 1;
            var rand = Math.floor(Math.random()*11);
            var property = Object.getOwnPropertyNames(root.db.perks)[rand];
            lvlArr.push(property);
            root.db.perks[property] += 1;
            root.players.update(root.db);
        }
    };
};
